### DISCUSSION TITLE
Best practices on using Next.js in a TypeScript monorepo

### DISCUSSION THREAD

**Discussion by PaulRBerg** [vercel/next.js#50866] [Help]
## Problem

I've spent a lot of time researching how to use Next.js v13 in TypeScript monorepo, and I haven't found much guidance. In fact, I've found conflicting recommendations.

Googling "TypeScript monorepo" leads to many articles[^1][^2] that recommend using TypeScript's [Project References](https://www.typescriptlang.org/docs/handbook/project-references.html) feature (PREFs), which involves enabling `composite` in `tsconfig.json`, as well as specifying all TypeScript dependencies in the `references` field. However, as @kirill-konshin pointed out [here](https://github.com/vercel/next.js/issues/8708#issue-492374452), PREFs are not compatible with Next.js (given that Next.js uses a different compiler, which at the moment is [SWC](https://nextjs.org/docs/architecture/nextjs-compiler)).

Conversely, googling "Next.js monorepo" leads to several different approaches. The two that stand out are:

- @belgattitude's [nextjs-monorepo-example](https://github.com/belgattitude/nextjs-monorepo-example), which employs a plurality of TSConfig files, some which handle building, and others which handle linting (for IDE support)
- Vercel Solutions' [monorepo example](https://github.com/vercel/examples/tree/9b2e5fbda9f7a972f35e0a2a5675bcedc0e3465b/solutions/monorepo), which uses [ES modules](https://github.com/vercel/examples/blob/9b2e5fbda9f7a972f35e0a2a5675bcedc0e3465b/solutions/monorepo/packages/utils/package.json#L22-L39) to provide selective type resolution

## Questions

Given the above, I have the following questions for the Vercel team:

1. Am I right that PREFs are off-the-table if one wishes to add a Next.js app that depends on other TypeScript packages?
    - Or, one could try to use PREFs in a _part_ of the monorepo that is Next.js-free, but one would then have a difficult time integrating that PREFs-enabled parts in the Next.js app, right? For instance, I can't think of a way to enable hot reloading in this context.
2. What would you recommend using between the two approaches mentioned above?
3. Generally speaking, how can one get the benefits of PREFs (modularity, separation of concerns, cached incremental builds, great debugging UX, etc.) in a Next.js context?

P.S. it seems to me that Next.js forces monorepo users to create a _dichotomy_ between (i) development-only TypeScript configs and (ii) production TypeScript configs. And I think it would be worth documenting this requirement in [the TypeScript section](https://nextjs.org/docs/pages/building-your-application/configuring/typescript) in your docs.

[^1]: https://dev.to/mxro/the-ultimate-guide-to-typescript-monorepos-5ap7
[^2]: https://earthly.dev/blog/setup-typescript-monorepo/

### Additional information

@Victordmdb asked the same question [here](https://github.com/vercel/next.js/issues/8708#issuecomment-537868136), but he has never received an answer.

**Comment by flippyhead** [comment]
FWIW the new monorepo support in ShadCN will create a new monorepo for a new NextJS project that has decent defaults.

**Reply by PaulRBerg** [reply]
Interesting. Could you share a link?

**Reply by Sumis34** [reply]
Is it this [monorepo-next](https://github.com/shadcn-ui/ui/tree/main/templates/monorepo-next)?
